File: benchmarks/55-vectorized-special-functions/vectorized_special_functions.py
#!/usr/bin/env python3
"""Experiment 55 -- Special functions over machine arrays (scipy.special column).

Same kernels as ``vectorized_special_functions.m``, same order and labels.

EXECUTION comparison: scipy.special is compiled C (SIMD ufuncs).  For the
vectorized Mathilda heads the gap is overhead; for BesselI/BesselK -- which have
no vector kernel in Mathilda and thread scalar-by-scalar -- the gap is a missing
kernel, and the ratio is large.  All ten are one ufunc call here.

Normalization: Mathematica's FresnelC/FresnelS and scipy.special.fresnel both
use the pi*t^2/2 argument; sici returns (Si, Ci); fresnel returns (S, C).
Checks are a rounded scalar at a fixed input.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
import scipy.special as sp

from harness import bench, check, require, bench_once, seed

require(["scipy.special:fresnel", "scipy.special:erfc", "scipy.special:erfi",
         "scipy.special:expi", "scipy.special:sici", "scipy.special:beta",
         "scipy.special:iv", "scipy.special:kv",
         "scipy.special:eval_legendre", "scipy.special:legendre_q"])

seed()
n = 1000000
v = np.random.random(n) + 0.5          # [0.5, 1.5]: safe for every function
w = v + 0.5
vc = np.random.random(n) - 0.5         # [-0.5, 0.5]: on the cut for LegendreQ


def r6(x):
    return int(np.floor(float(x) * 1e6 + 0.5))


def legendre_q_np(deg, x):
    """Vectorized Q_n on the cut via the 3-term recurrence (scipy has no
    vectorized legendre_q; only the scalar lqn), so this numpy recurrence is
    what a scipy user writes.  Q0 = arctanh(x), Q1 = x*Q0 - 1."""
    q0 = np.arctanh(x)
    if deg == 0:
        return q0
    q1 = x * q0 - 1.0
    for k in range(2, deg + 1):
        q0, q1 = q1, ((2 * k - 1) * x * q1 - (k - 1) * q0) / k
    return q1


# ---- vectorized on both sides: fair overhead comparison -----------------

bench("FresnelC over 10^6", lambda: sp.fresnel(v)[1])
check("FresnelC over 10^6", r6(sp.fresnel(0.5)[1]))

bench("FresnelS over 10^6", lambda: sp.fresnel(v)[0])
check("FresnelS over 10^6", r6(sp.fresnel(0.5)[0]))

bench("Erfc over 10^6", lambda: sp.erfc(v))
check("Erfc over 10^6", r6(sp.erfc(0.5)))

bench("Erfi over 10^6", lambda: sp.erfi(v))
check("Erfi over 10^6", r6(sp.erfi(0.5)))

bench("ExpIntegralEi over 10^6", lambda: sp.expi(v))
check("ExpIntegralEi over 10^6", r6(sp.expi(1.5)))

bench("SinIntegral over 10^6", lambda: sp.sici(v)[0])
check("SinIntegral over 10^6", r6(sp.sici(1.5)[0]))

bench("CosIntegral over 10^6", lambda: sp.sici(v)[1])
check("CosIntegral over 10^6", r6(sp.sici(1.5)[1]))

bench("Beta[.,.] over 10^6", lambda: sp.beta(v, w))
check("Beta[.,.] over 10^6", r6(sp.beta(1.5, 2.0)))

# ---- Legendre functions of both kinds over the cut ----------------------
# P: scipy.special.eval_legendre is a compiled vectorized ufunc (legendre_p
# exists too but returns a stacked shape).  Q: scipy has NO vectorized second
# kind, so the numpy recurrence above stands in.

bench("LegendreP[10, .] over 10^6", lambda: sp.eval_legendre(10, vc))
check("LegendreP[10, .] over 10^6", r6(sp.eval_legendre(10, 0.5)))

bench("LegendreQ[10, .] over 10^6", lambda: legendre_q_np(10, vc))
check("LegendreQ[10, .] over 10^6", r6(legendre_q_np(10, 0.5)))

# ---- scipy vectorized; Mathilda scalar-threaded (missing kernel) --------

bench_once("BesselI[0, .] over 10^6", lambda: sp.iv(0, v))
check("BesselI[0, .] over 10^6", r6(sp.iv(0, 1.5)))

bench_once("BesselK[0, .] over 10^6", lambda: sp.kv(0, v))
check("BesselK[0, .] over 10^6", r6(sp.kv(0, 1.5)))

File: benchmarks/55-vectorized-special-functions/vectorized_special_functions.m
(* ==========================================================================
   Experiment 55 -- Special functions over machine arrays (group E)
   ==========================================================================
   WHAT IT MEASURES.  The elementwise special-function surface reached through a
   10^6 packed buffer -- src/ndkernels.c into src/special_functions/ -- the
   direct analogue of a scipy.special ufunc.  Experiment 11 covers
   Gamma/Erf/BesselJ/LogGamma/PolyGamma/AiryAi/Zeta; this covers the rest of the
   library with a scipy.special counterpart, and in doing so answers a sharper
   question: WHICH heads have a SIMD vector kernel and which fall back to
   scalar-by-scalar threading.

   THE FINDING.  The Fresnel integrals, the complementary/imaginary error
   functions, the exponential/trigonometric integrals and the two-argument Beta
   are vectorized -- tens of ms over 10^6, a fair overhead comparison against
   scipy's compiled ufuncs.  The modified Bessel functions BesselI/BesselK now
   also run their registered NDArray kernel (tens of ms over 10^6, faster than
   scipy.special.iv/kv): the transparency gate used to force any rule-bearing
   head -- and every Bessel head carries half-integer DownValues -- off the
   buffer path, so they threaded one element at a time (~20-40 s).  The gate now
   also admits a head with a matching-arity NDArray kernel (eval.c), with the
   .m rules guarded by !NDArrayQ so they decline on a buffer.  benchOnce is kept
   so the row stays comparable across the history.

   LEGENDRE, BOTH KINDS.  LegendreP/LegendreQ also carry NDArray binary kernels
   (sf_machine_legendre_p/q): a degree-10 three-term recurrence per element on a
   packed 10^6 buffer, a few ms each.  LegendreP beats scipy.special.eval_legendre
   on wall-clock; LegendreQ has NO vectorized scipy counterpart at all (only the
   scalar lqn, ~1000x slower), so its .py baseline is a numpy recurrence.

   Checks are a rounded scalar at a fixed rational input, never a sum over the
   random timing data.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"FresnelC", "FresnelS", "Erfc", "Erfi", "ExpIntegralEi",
         "SinIntegral", "CosIntegral", "Beta", "BesselI", "BesselK",
         "LegendreP", "LegendreQ"}];

seed[];
n = 1000000;
v = rand01[{n}] + 1/2;              (* in [0.5, 1.5]: safe for every function *)
w = v + 1/2;
vc = rand01[{n}] - 1/2;             (* in [-0.5, 0.5]: on the cut for LegendreQ *)

(* ---- vectorized: buffer-speed, fair overhead comparison ---------------- *)

bench["FresnelC over 10^6", FresnelC[v];];
check["FresnelC over 10^6", Round[10^6 N[FresnelC[1/2]]]];

bench["FresnelS over 10^6", FresnelS[v];];
check["FresnelS over 10^6", Round[10^6 N[FresnelS[1/2]]]];

bench["Erfc over 10^6", Erfc[v];];
check["Erfc over 10^6", Round[10^6 N[Erfc[1/2]]]];

bench["Erfi over 10^6", Erfi[v];];
check["Erfi over 10^6", Round[10^6 N[Erfi[1/2]]]];

bench["ExpIntegralEi over 10^6", ExpIntegralEi[v];];
check["ExpIntegralEi over 10^6", Round[10^6 N[ExpIntegralEi[3/2]]]];

bench["SinIntegral over 10^6", SinIntegral[v];];
check["SinIntegral over 10^6", Round[10^6 N[SinIntegral[3/2]]]];

bench["CosIntegral over 10^6", CosIntegral[v];];
check["CosIntegral over 10^6", Round[10^6 N[CosIntegral[3/2]]]];

bench["Beta[.,.] over 10^6", Beta[v, w];];
check["Beta[.,.] over 10^6", Round[10^6 N[Beta[3/2, 2]]]];

(* ---- Legendre functions of both kinds over the cut -------------------- *)
(* Both run their registered NDArray binary kernel (sf_machine_legendre_p/q via
   src/ndkernels.c): a degree-10 three-term recurrence per element on a packed
   10^6 buffer.  LegendreQ has NO scipy counterpart -- the .py side times a
   numpy recurrence -- so its column is a kernel-vs-numpy comparison. *)

bench["LegendreP[10, .] over 10^6", LegendreP[10, vc];];
check["LegendreP[10, .] over 10^6", Round[10^6 N[LegendreP[10, 1/2]]]];

bench["LegendreQ[10, .] over 10^6", LegendreQ[10, vc];];
check["LegendreQ[10, .] over 10^6", Round[10^6 N[LegendreQ[10, 1/2]]]];

(* ---- no vector kernel: scalar threading, benchOnce --------------------- *)

benchOnce["BesselI[0, .] over 10^6", BesselI[0, v];];
check["BesselI[0, .] over 10^6", Round[10^6 N[BesselI[0, 3/2]]]];

benchOnce["BesselK[0, .] over 10^6", BesselK[0, v];];
check["BesselK[0, .] over 10^6", Round[10^6 N[BesselK[0, 3/2]]]];
